src/main/java/com/microsoft/azure/relay/AsyncSemaphore.java

```diff
@@ -89,6 +91,10 @@ int availablePermits() {
 		});
 	}
 	
+	public <T> CompletableFuture<T> runInsideLockAsync(Duration timeout, AutoShutdownScheduledExecutor executor, Supplier<? extends CompletionStage<T>> supplier) {
```

Comment:
>AutoShutdownScheduledExecutor [](start = 70, length = 29)

Maybe use ExecutorService instead here?

Reply:
I'll look into it. The Problem is what type does AsyncSemaphore.acquireAsync take?

---
In reply to: [259064026](https://github.com/Azure/azure-relay-java/pull/20#discussion_r259064026) [](ancestors = 259064026)

Reply:
I'd have to change the parameters for AsyncSemaphore and InputQueue. Whee!

---
In reply to: [259476726](https://github.com/Azure/azure-relay-java/pull/20#discussion_r259476726) [](ancestors = 259476726,259064026)

Reply:
That's fine then, not a big deal.

---
In reply to: [259477433](https://github.com/Azure/azure-relay-java/pull/20#discussion_r259477433) [](ancestors = 259477433,259476726,259064026)